Index: libressl/include/openssl/Makefile.am
===================================================================
--- libressl.orig/include/openssl/Makefile.am	2015-03-05 15:11:46.000000000 +0100
+++ libressl/include/openssl/Makefile.am	2015-03-05 15:17:26.000000000 +0100
@@ -1,6 +1,6 @@
 include $(top_srcdir)/Makefile.am.common
 
-opensslincludedir=$(includedir)/openssl
+opensslincludedir=$(includedir)/libressl/openssl
 
 opensslinclude_HEADERS =
 opensslinclude_HEADERS += aes.h
Index: libressl/libressl-crypto.pc.in
===================================================================
--- libressl.orig/libressl-crypto.pc.in	2015-03-05 15:13:22.000000000 +0100
+++ libressl/libressl-crypto.pc.in	2015-03-05 15:17:26.000000000 +0100
@@ -12,4 +12,4 @@
 Conflicts:
 Libs: -L${libdir} -lressl-crypto
 Libs.private: @LIBS@
-Cflags: -I${includedir}
+Cflags: -I${includedir}/libressl
Index: libressl/libressl-ssl.pc.in
===================================================================
--- libressl.orig/libressl-ssl.pc.in	2015-03-05 15:13:22.000000000 +0100
+++ libressl/libressl-ssl.pc.in	2015-03-05 15:17:26.000000000 +0100
@@ -13,4 +13,4 @@
 Conflicts:
 Libs: -L${libdir} -lressl-ssl
 Libs.private: @LIBS@ -lressl-crypto
-Cflags: -I${includedir}
+Cflags: -I${includedir}/libressl
Index: libressl/libressl-tls.pc.in
===================================================================
--- libressl.orig/libressl-tls.pc.in	2015-03-05 15:13:22.000000000 +0100
+++ libressl/libressl-tls.pc.in	2015-03-05 15:17:26.000000000 +0100
@@ -13,4 +13,4 @@
 Conflicts:
 Libs: -L${libdir} -lressl-tls
 Libs.private: @LIBS@ -lressl-crypto -lressl-ssl
-Cflags: -I${includedir}
+Cflags: -I${includedir}/libressl
Index: libressl/include/Makefile.am
===================================================================
--- libressl.orig/include/Makefile.am	2015-03-05 15:11:46.000000000 +0100
+++ libressl/include/Makefile.am	2015-03-05 15:17:26.000000000 +0100
@@ -29,5 +29,6 @@
 noinst_HEADERS += sys/uio.h
 
 if ENABLE_LIBTLS
-include_HEADERS = tls.h
+tlsincludedir=$(includedir)/libressl
+tlsinclude_HEADERS = tls.h
 endif
