--- a/include/openssl/Makefile.am
+++ b/include/openssl/Makefile.am
@@ -1,6 +1,6 @@
 include $(top_srcdir)/Makefile.am.common
 
-opensslincludedir=$(includedir)/openssl
+opensslincludedir=$(includedir)/libressl/openssl
 
 opensslinclude_HEADERS =
 opensslinclude_HEADERS += aes.h
--- a/libressl-crypto.pc.in
+++ b/libressl-crypto.pc.in
@@ -12,4 +12,4 @@
 Conflicts:
 Libs: -L${libdir} -lressl-crypto
 Libs.private: @LIBS@
-Cflags: -I${includedir}
+Cflags: -I${includedir}/libressl
--- a/libressl-ssl.pc.in
+++ b/libressl-ssl.pc.in
@@ -13,4 +13,4 @@
 Conflicts:
 Libs: -L${libdir} -lressl-ssl
 Libs.private: @LIBS@ -lressl-crypto
-Cflags: -I${includedir}
+Cflags: -I${includedir}/libressl
--- a/libressl-tls.pc.in
+++ b/libressl-tls.pc.in
@@ -13,4 +13,4 @@
 Conflicts:
 Libs: -L${libdir} -lressl-tls
 Libs.private: @LIBS@ -lressl-crypto -lressl-ssl
-Cflags: -I${includedir}
+Cflags: -I${includedir}/libressl
--- a/include/Makefile.am
+++ b/include/Makefile.am
@@ -40,4 +40,5 @@
 noinst_HEADERS += compat/sys/types.h
 noinst_HEADERS += compat/sys/uio.h
 
-include_HEADERS = tls.h
+tlsincludedir=$(includedir)/libressl
+tlsinclude_HEADERS = tls.h
